File: extensions/claude_stream.py
"""Parse Claude CLI stream-json output into structured harness events."""

from __future__ import annotations

import json
from typing import Any, Generator, Iterator


def parse_claude_stream(lines: Iterator[str]) -> Generator[dict[str, Any], None, None]:
    parser = _ClaudeStreamParser()
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            envelope = json.loads(line)
        except json.JSONDecodeError:
            continue
        yield from parser.handle_envelope(envelope)


class _ClaudeStreamParser:
    def __init__(self) -> None:
        self.emitted_text = False
        self.seen_tool_starts: set[str] = set()
        self.seen_tool_ends: set[str] = set()
        self.seen_tool_results: set[str] = set()
        self.blocks: dict[int, dict[str, Any]] = {}

    def handle_envelope(self, envelope: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        t = envelope.get("type")
        if t == "stream_event":
            yield from self._handle_stream_event(envelope.get("event") or {})
        elif t == "assistant":
            yield from self._handle_assistant((envelope.get("message") or {}).get("content") or [])
        elif t == "user":
            yield from self._handle_user(envelope)
        elif t == "result":
            if envelope.get("is_error"):
                yield {"type": "error", "error": envelope.get("error") or "unknown error"}
            else:
                sid = envelope.get("session_id") or ""
                if sid:
                    yield {"type": "session_id", "sessionId": sid}

    def _handle_stream_event(self, ev: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        ev_type = ev.get("type")
        if ev_type == "content_block_delta":
            delta = ev.get("delta") or {}
            if delta.get("type") == "text_delta" and delta.get("text"):
                self.emitted_text = True
                yield {"type": "text", "content": delta["text"]}
            elif delta.get("type") == "input_json_delta" and delta.get("partial_json"):
                idx = ev.get("index")
                state = self.blocks.get(idx)
                if state and state.get("kind") == "tool_use":
                    state["args"] += delta["partial_json"]
        elif ev_type == "content_block_start":
            block = ev.get("content_block") or {}
            if block.get("type") != "tool_use" or not block.get("id"):
                return
            idx = ev.get("index")
            self.blocks[idx] = {
                "kind": "tool_use",
                "tool_id": block["id"],
                "tool_name": block.get("name") or "",
                "args": "",
            }
            if block["id"] in self.seen_tool_starts:
                return
            self.seen_tool_starts.add(block["id"])
            yield {
                "type": "tool_call_start",
                "toolCallId": block["id"],
                "toolName": block.get("name") or "",
            }
        elif ev_type == "content_block_stop":
            idx = ev.get("index")
            state = self.blocks.pop(idx, None)
            if not state or state.get("kind") != "tool_use":
                return
            args = state.get("args") or ""
            if args:
                yield {
                    "type": "tool_call_args",
                    "toolCallId": state["tool_id"],
                    "toolArgs": args,
                }
            if state["tool_id"] not in self.seen_tool_ends:
                self.seen_tool_ends.add(state["tool_id"])
                yield {
                    "type": "tool_call_end",
                    "toolCallId": state["tool_id"],
                    "toolName": state["tool_name"],
                    "toolArgs": args,
                }

    def _handle_assistant(self, blocks: list[Any]) -> Generator[dict[str, Any], None, None]:
        for block in blocks:
            if not isinstance(block, dict):
                continue
            btype = block.get("type")
            if btype == "text" and block.get("text") and not self.emitted_text:
                yield {"type": "text", "content": block["text"]}
            elif btype == "tool_use" and block.get("id"):
                tool_id = block["id"]
                args = json.dumps(block.get("input") or {})
                if tool_id not in self.seen_tool_starts:
                    self.seen_tool_starts.add(tool_id)
                    yield {
                        "type": "tool_call_start",
                        "toolCallId": tool_id,
                        "toolName": block.get("name") or "",
                    }
                    yield {"type": "tool_call_args", "toolCallId": tool_id, "toolArgs": args}
                if tool_id not in self.seen_tool_ends:
                    self.seen_tool_ends.add(tool_id)
                    yield {
                        "type": "tool_call_end",
                        "toolCallId": tool_id,
                        "toolName": block.get("name") or "",
                        "toolArgs": args,
                    }
            elif btype == "image":
                md = _image_block_markdown(block)
                if md:
                    yield {"type": "text", "content": md}

    def _handle_user(self, envelope: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        tool_use_id = envelope.get("parent_tool_use_id") or ""
        tool_use_result = envelope.get("tool_use_result")
        message = envelope.get("message") or {}
        content = message.get("content")

        if not tool_use_id and isinstance(content, list):
            for block in content:
                if isinstance(block, dict) and block.get("type") == "tool_result":
                    tool_use_id = block.get("tool_use_id") or tool_use_id

        if tool_use_id and tool_use_result is not None:
            yield from self._emit_tool_result(tool_use_id, tool_use_result)

        if isinstance(content, list):
            for block in content:
                if not isinstance(block, dict) or block.get("type") != "tool_result":
                    continue
                tid = block.get("tool_use_id") or ""
                if not tid:
                    continue
                yield from self._emit_tool_result(tid, block.get("content"))
                for md in _image_content_markdown(block.get("content")):
                    yield {"type": "text", "content": md}

    def _emit_tool_result(self, tool_use_id: str, result: Any) -> Generator[dict[str, Any], None, None]:
        if tool_use_id in self.seen_tool_results:
            return
        self.seen_tool_results.add(tool_use_id)
        if isinstance(result, str):
            content = result
        else:
            content = json.dumps(result)
        yield {
            "type": "tool_call_result",
            "toolCallId": tool_use_id,
            "content": content,
        }


def _image_block_markdown(block: dict[str, Any]) -> str:
    source = block.get("source") or {}
    return _image_source_markdown(
        source.get("type") or "",
        source.get("media_type") or "",
        source.get("data") or "",
        source.get("url") or "",
    )


def _image_content_markdown(content: Any) -> list[str]:
    out: list[str] = []
    if isinstance(content, list):
        for block in content:
            if isinstance(block, dict) and block.get("type") == "image":
                md = _image_block_markdown(block)
                if md:
                    out.append(md)
        return out
    if isinstance(content, dict):
        _walk_images(content, out)
    return out


def _walk_images(value: Any, out: list[str]) -> None:
    if isinstance(value, dict):
        if value.get("type") == "image":
            source = value.get("source") or {}
            md = _image_source_markdown(
                source.get("type") or "",
                source.get("media_type") or "",
                source.get("data") or "",
                source.get("url") or "",
            )
            if md:
                out.append(md)
        for child in value.values():
            _walk_images(child, out)
    elif isinstance(value, list):
        for child in value:
            _walk_images(child, out)


def _image_source_markdown(src_type: str, media_type: str, data: str, url: str) -> str:
    if src_type == "base64" and data:
        media_type = media_type or "image/png"
        return f"\n\n![image](data:{media_type};base64,{data})\n\n"
    if url:
        return f"\n\n![image]({url})\n\n"
    return ""

File: internal/agent/claude_stream.go
// Copyright (c) Mehmet Bektas <mbektasgh@outlook.com>

package agent

import (
	"encoding/json"
	"strings"
)

type claudeStreamParser struct {
	emittedText     bool
	needsTextSep    bool
	seenToolStarts  map[string]bool
	seenToolEnds    map[string]bool
	seenToolResults map[string]bool
	blocks          map[int]*claudeBlockState
	sessionID       string
	lastStopReason  string
	emittedDone     bool
	turnComplete    bool
}

type claudeBlockState struct {
	kind     string
	toolID   string
	toolName string
	args     strings.Builder
}

func newClaudeStreamParser() *claudeStreamParser {
	return &claudeStreamParser{
		seenToolStarts:  map[string]bool{},
		seenToolEnds:    map[string]bool{},
		seenToolResults: map[string]bool{},
		blocks:          map[int]*claudeBlockState{},
	}
}

func (p *claudeStreamParser) emitDone(sessionID, parentToolUseID string, events chan<- Event) {
	if parentToolUseID != "" {
		return
	}
	if p.emittedDone || p.hasPendingToolWork() {
		return
	}
	p.emittedDone = true
	p.turnComplete = true
	sid := sessionID
	if sid == "" {
		sid = p.sessionID
	}
	events <- Event{Type: EventDone, SessionID: sid}
}

func (p *claudeStreamParser) hasPendingToolWork() bool {
	for id := range p.seenToolStarts {
		if !p.seenToolResults[id] {
			return true
		}
	}
	return false
}

func (p *claudeStreamParser) completedTurn() bool {
	return p.turnComplete
}

func (p *claudeStreamParser) markTextSepNeeded() {
	if p.emittedText {
		p.needsTextSep = true
	}
}

func (p *claudeStreamParser) emit(parentToolUseID string, ev Event, events chan<- Event) {
	if parentToolUseID != "" {
		ev.ParentToolCallID = parentToolUseID
	}
	events <- ev
}

func (p *claudeStreamParser) emitText(text, parentToolUseID string, events chan<- Event) {
	if text == "" {
		return
	}
	if p.emittedText && p.needsTextSep && !strings.HasPrefix(text, "\n") {
		text = "\n\n" + text
	}
	p.needsTextSep = false
	p.emittedText = true
	p.emit(parentToolUseID, Event{Type: EventText, Content: text}, events)
}

func (p *claudeStreamParser) handleLine(line []byte, events chan<- Event) {
	if len(line) == 0 {
		return
	}

	var envelope struct {
		Type             string          `json:"type"`
		Event            json.RawMessage `json:"event"`
		SessionID        string          `json:"session_id"`
		IsError          bool            `json:"is_error"`
		ErrMsg           string          `json:"error"`
		Result           string          `json:"result"`
		ParentToolUseID  string          `json:"parent_tool_use_id"`
		ToolUseResult    json.RawMessage `json:"tool_use_result"`
		Message          json.RawMessage `json:"message"`
	}
	if err := json.Unmarshal(line, &envelope); err != nil {
		return
	}

	if envelope.SessionID != "" {
		p.sessionID = envelope.SessionID
	}

	parentToolUseID := envelope.ParentToolUseID

	switch envelope.Type {
	case "stream_event":
		p.handleStreamEvent(envelope.Event, parentToolUseID, events)
	case "assistant":
		p.handleAssistant(envelope.Message, parentToolUseID, events)
	case "user":
		p.handleUser(parentToolUseID, envelope.ToolUseResult, envelope.Message, events)
	case "result":
		if envelope.IsError {
			msg := envelope.ErrMsg
			if msg == "" {
				msg = envelope.Result
			}
			p.emit(parentToolUseID, Event{Type: EventError, Error: msg}, events)
		} else {
			p.emitDone(envelope.SessionID, parentToolUseID, events)
		}
	}
}

func (p *claudeStreamParser) handleStreamEvent(raw json.RawMessage, parentToolUseID string, events chan<- Event) {
	var ev struct {
		Type  string `json:"type"`
		Index int    `json:"index"`
		Delta struct {
			Type        string `json:"type"`
			Text        string `json:"text"`
			PartialJSON string `json:"partial_json"`
			StopReason  string `json:"stop_reason"`
		} `json:"delta"`
		ContentBlock struct {
			Type string `json:"type"`
			ID   string `json:"id"`
			Name string `json:"name"`
		} `json:"content_block"`
	}
	if err := json.Unmarshal(raw, &ev); err != nil {
		return
	}

	switch ev.Type {
	case "message_delta":
		if ev.Delta.StopReason != "" {
			p.lastStopReason = ev.Delta.StopReason
		}
	case "message_stop":
		if ev.Delta.StopReason != "" {
			p.lastStopReason = ev.Delta.StopReason
		}
		if p.lastStopReason == "end_turn" {
			p.emitDone("", parentToolUseID, events)
		}
		p.lastStopReason = ""
	case "content_block_delta":
		if ev.Delta.Type == "text_delta" && ev.Delta.Text != "" {
			p.emitText(ev.Delta.Text, parentToolUseID, events)
		}
	case "content_block_start":
		block := ev.ContentBlock
		if block.Type == "text" {
			p.markTextSepNeeded()
			return
		}
		if block.Type != "tool_use" || block.ID == "" {
			return
		}
		p.markTextSepNeeded()
		p.blocks[ev.Index] = &claudeBlockState{
			kind:     "tool_use",
			toolID:   block.ID,
			toolName: block.Name,
		}
		if parentToolUseID == "" {
			if p.seenToolStarts[block.ID] {
				return
			}
			p.seenToolStarts[block.ID] = true
		}
		p.emit(parentToolUseID, Event{
			Type:       EventToolCallStart,
			ToolCallID: block.ID,
			ToolName:   block.Name,
		}, events)
	case "content_block_stop":
		state, ok := p.blocks[ev.Index]
		if !ok || state.kind != "tool_use" {
			delete(p.blocks, ev.Index)
			return
		}
		args := state.args.String()
		if args != "" {
			p.emit(parentToolUseID, Event{
				Type:       EventToolCallArgs,
				ToolCallID: state.toolID,
				ToolArgs:   args,
			}, events)
		}
		if parentToolUseID == "" {
			if !p.seenToolEnds[state.toolID] {
				p.seenToolEnds[state.toolID] = true
				p.emit(parentToolUseID, Event{
					Type:       EventToolCallEnd,
					ToolCallID: state.toolID,
					ToolName:   state.toolName,
					ToolArgs:   args,
				}, events)
			}
		} else {
			scopeKey := parentToolUseID + "::" + state.toolID
			if !p.seenToolEnds[scopeKey] {
				p.seenToolEnds[scopeKey] = true
				p.emit(parentToolUseID, Event{
					Type:       EventToolCallEnd,
					ToolCallID: state.toolID,
					ToolName:   state.toolName,
					ToolArgs:   args,
				}, events)
			}
		}
		p.markTextSepNeeded()
		delete(p.blocks, ev.Index)
	}

	if ev.Type == "content_block_delta" && ev.Delta.Type == "input_json_delta" && ev.Delta.PartialJSON != "" {
		if state, ok := p.blocks[ev.Index]; ok && state.kind == "tool_use" {
			state.args.WriteString(ev.Delta.PartialJSON)
		}
	}
}

func (p *claudeStreamParser) handleAssistant(raw json.RawMessage, parentToolUseID string, events chan<- Event) {
	var msg struct {
		Content []json.RawMessage `json:"content"`
	}
	if err := json.Unmarshal(raw, &msg); err != nil {
		return
	}

	for _, blockRaw := range msg.Content {
		var blockType struct {
			Type string `json:"type"`
		}
		if err := json.Unmarshal(blockRaw, &blockType); err != nil {
			continue
		}

		switch blockType.Type {
		case "text":
			var block struct {
				Text string `json:"text"`
			}
			if json.Unmarshal(blockRaw, &block) != nil || block.Text == "" {
				continue
			}
			// Partial stream_event deltas are authoritative with --include-partial-messages.
			// Full assistant snapshots after tool/HITL pauses can replay prior turns.
			if parentToolUseID == "" && (len(p.seenToolStarts) > 0 || p.emittedText) {
				continue
			}
			p.emitText(block.Text, parentToolUseID, events)
		case "tool_use":
			p.markTextSepNeeded()
			var block struct {
				ID    string         `json:"id"`
				Name  string         `json:"name"`
				Input map[string]any `json:"input"`
			}
			if err := json.Unmarshal(blockRaw, &block); err != nil || block.ID == "" {
				continue
			}
			argsJSON, _ := json.Marshal(block.Input)
			argsStr := string(argsJSON)
			if parentToolUseID == "" {
				if !p.seenToolStarts[block.ID] {
					p.seenToolStarts[block.ID] = true
					p.emit(parentToolUseID, Event{
						Type:       EventToolCallStart,
						ToolCallID: block.ID,
						ToolName:   block.Name,
					}, events)
					p.emit(parentToolUseID, Event{
						Type:       EventToolCallArgs,
						ToolCallID: block.ID,
						ToolArgs:   argsStr,
					}, events)
				}
				if !p.seenToolEnds[block.ID] {
					p.seenToolEnds[block.ID] = true
					p.emit(parentToolUseID, Event{
						Type:       EventToolCallEnd,
						ToolCallID: block.ID,
						ToolName:   block.Name,
						ToolArgs:   argsStr,
					}, events)
				}
			} else {
				p.emit(parentToolUseID, Event{
					Type:       EventToolCallStart,
					ToolCallID: block.ID,
					ToolName:   block.Name,
				}, events)
				p.emit(parentToolUseID, Event{
					Type:       EventToolCallArgs,
					ToolCallID: block.ID,
					ToolArgs:   argsStr,
				}, events)
				p.emit(parentToolUseID, Event{
					Type:       EventToolCallEnd,
					ToolCallID: block.ID,
					ToolName:   block.Name,
					ToolArgs:   argsStr,
				}, events)
			}
		case "image":
			emitImageEvents(blockRaw, parentToolUseID, events)
		}
	}
}

func (p *claudeStreamParser) handleUser(parentToolUseID string, toolUseResult json.RawMessage, messageRaw json.RawMessage, events chan<- Event) {
	toolUseID := parentToolUseID
	if toolUseID == "" {
		toolUseID = toolUseIDFromMessage(messageRaw)
	}
	if toolUseID != "" && len(toolUseResult) > 0 && toolUseResult[0] != 'n' {
		p.emitToolResult(toolUseID, "", toolUseResult, events)
		emitImageEvents(toolUseResult, "", events)
	}

	var msg struct {
		Content json.RawMessage `json:"content"`
	}
	if err := json.Unmarshal(messageRaw, &msg); err != nil {
		return
	}
	var blocks []json.RawMessage
	if err := json.Unmarshal(msg.Content, &blocks); err != nil {
		return
	}
	for _, blockRaw := range blocks {
		var block struct {
			Type      string          `json:"type"`
			ToolUseID string          `json:"tool_use_id"`
			Content   json.RawMessage `json:"content"`
		}
		if err := json.Unmarshal(blockRaw, &block); err != nil {
			continue
		}
		if block.Type != "tool_result" || block.ToolUseID == "" {
			continue
		}
		p.emitToolResult(block.ToolUseID, parentToolUseID, block.Content, events)
		emitImageEvents(block.Content, parentToolUseID, events)
	}
}

func (p *claudeStreamParser) emitToolResult(toolUseID, parentToolUseID string, result json.RawMessage, events chan<- Event) {
	if parentToolUseID == "" {
		if p.seenToolResults[toolUseID] {
			return
		}
		p.seenToolResults[toolUseID] = true
	} else if p.seenToolResults[parentToolUseID+"::"+toolUseID] {
		return
	} else {
		p.seenToolResults[parentToolUseID+"::"+toolUseID] = true
	}
	p.emit(parentToolUseID, Event{
		Type:       EventToolCallResult,
		ToolCallID: toolUseID,
		Content:    string(result),
	}, events)
}

func toolUseIDFromMessage(messageRaw json.RawMessage) string {
	var msg struct {
		Content json.RawMessage `json:"content"`
	}
	if err := json.Unmarshal(messageRaw, &msg); err != nil {
		return ""
	}
	var blocks []json.RawMessage
	if err := json.Unmarshal(msg.Content, &blocks); err != nil {
		return ""
	}
	for _, blockRaw := range blocks {
		var block struct {
			Type      string `json:"type"`
			ToolUseID string `json:"tool_use_id"`
		}
		if json.Unmarshal(blockRaw, &block) == nil && block.Type == "tool_result" && block.ToolUseID != "" {
			return block.ToolUseID
		}
	}
	return ""
}
